bigframes/core/__init__.py

```diff
@@ -49,470 +40,196 @@
-                new_type: ibis_dtypes.DataType = ibis_dtypes.int64
-            else:
-                column_type = schema[column]
-                # The autodetected type might not be one we can support, such
-                # as NULL type for empty rows, so convert to a type we do
-                # support.
-                new_type = bigframes.dtypes.bigframes_dtype_to_ibis_dtype(
-                    bigframes.dtypes.ibis_dtype_to_bigframes_dtype(column_type)
-                )
-                # TODO(swast): Ibis memtable doesn't use backticks in struct
-                # field names, so spaces and other characters aren't allowed in
-                # the memtable context. Blocked by
-                # https://github.com/ibis-project/ibis/issues/7187
-                column = f"col_{column_index}"
-            new_schema.append((column, new_type))
+    def compile(self) -> compiled.CompiledArrayValue:
+        return compiled.ArrayValueCompiler().compile(self)
 
-        # must set non-null column labels. these are not the user-facing labels
-        pd_df = pd_df.set_axis(
-            [column for column, _ in new_schema],
-            axis="columns",
-        )
-        keys_memtable = ibis.memtable(pd_df, schema=ibis.schema(new_schema))
+    def shape(self) -> typing.Tuple[int, int]:
+        """Returns dimensions as (length, width) tuple."""
+        width = len(self.compile().columns)
+        count_expr = self.compile()._to_ibis_expr("unordered").count()
 
-        return cls(
-            session,  # type: ignore # Session cannot normally be none, see "caution" above
-            keys_memtable,
-            columns=[
-                keys_memtable[f"col_{column_index}"].name(column)
-                for column_index, column in enumerate(column_names)
-            ],
-            ordering=ExpressionOrdering(
-                ordering_value_columns=[OrderingColumnReference(ORDER_ID_COLUMN)],
-                total_ordering_columns=frozenset([ORDER_ID_COLUMN]),
-            ),
-            hidden_ordering_columns=(keys_memtable[ORDER_ID_COLUMN],),
+        # Support in-memory engines for hermetic unit tests.
+        if not self.node.sessions:
+            try:
+                length = ibis.pandas.connect({}).execute(count_expr)
+                return (length, width)
+            except Exception:
+                # Not all cases can be handled by pandas engine
+                pass
+
+        sql = self.session.ibis_client.compile(count_expr)
+        row_iterator, _ = self.session._start_query(
+            sql=sql,
+            max_results=1,
         )
+        length = next(row_iterator)[0]
+        return (length, width)
 
-    @property
-    def columns(self) -> typing.Tuple[ibis_types.Value, ...]:
-        return self._columns
-
-    @property
-    def column_ids(self) -> typing.Sequence[str]:
-        return tuple(self._column_names.keys())
+    def _uniform_sampling(self, fraction: float) -> ArrayValue:
```

Comment:
TODO: move to be by the other node operators. (Comment separating sections)

Reply:
done